### PR TITLE
Store GUIConfiguration.WebauthnUserId as raw binary

### PR DESCRIPTION
**Pull request by emlun**
This is a meta-PR into:
- https://github.com/syncthing/syncthing/pull/9175

This eliminates a possible decoding error in `api.newWebauthnService`.

When we generate a random ID we hex encode it, so that the raw binary values look nice in `config.xml`, but we use the resulting hex encoding as the raw value and don't actually parse the hex.

@imsodin This should address https://github.com/syncthing/syncthing/pull/9175#discussion_r1712631956.

### CODE REVIEW

File: lib/api/api.go [hunk 1]
```diff
@@ -497,7 +498,7 @@ func (s *service) CommitConfiguration(from, to config.Configuration) bool {
 	// No action required when this changes, so mask the fact that it changed at all.
 	from.GUI.Debugging = to.GUI.Debugging
 
-	if to.GUI == from.GUI {
+	if cmp.Equal(to.GUI, from.GUI) {
```
Comment (imsodin) [inline lib/api/api.go]: What's the reason for this change?

Had a quick look at it's doc, and this seems like a dealbreaker:

> If no custom equality functions are used and no Equal method is defined, equality is determined by recursively comparing the primitive kinds on both values, much like [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual). Unlike [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual), unexported fields are not compared by default; they result in panics unless suppressed by using an [Ignore](https://pkg.go.dev/github.com/google/go-cmp/cmp#Ignore) option (see [github.com/google/go-cmp/cmp/cmpopts.IgnoreUnexported](https://pkg.go.dev/github.com/google/go-cmp@v0.6.0/cmp/cmpopts#IgnoreUnexported)) or explicitly compared using the [Exporter](https://pkg.go.dev/github.com/google/go-cmp/cmp#Exporter) option.

Probably there will newer be unexported fields, but it's not impossible, and we definitely don't want panics if it ever happens.

Also on a second look:

> It is intended to only be used in tests, [...]

**Review by imsodin — APPROVED** [review]
Looks good. Except for the `cmp` introduction. Approved anyway as the main parts work, if cmp makes it into the "main PR" I'll just pont it out there again :)

File: lib/api/api.go (hunk 1 above)
Comment (emlun) [reply lib/api/api.go]: Oh, my bad. The reason is this:

```
lib/api/api.go:501:5: invalid operation: to.GUI == from.GUI (struct containing []byte cannot be compared)
```

So _something_ else than `==` is needed for this to work. Is [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual) the correct solution to this, then?

Comment (emlun) [reply lib/api/api.go]: (This also snuck into the main PR in https://github.com/syncthing/syncthing/pull/9175/commits/844516eca2735051a3faec04c529f35c4e29d08f, but I'll amend that too with whatever resolution we reach here)

Comment (emlun) [reply lib/api/api.go]: Perhaps this part of [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual) could be problematic:

>Note that a non-nil empty slice and a nil slice (for example, []byte{} and []byte(nil)) are not deeply equal.

But it seems to at least pass all the tests both in this PR and in https://github.com/syncthing/syncthing/pull/9175.

File: lib/config/guiconfiguration.go [hunk 2]
```diff
@@ -166,13 +166,15 @@ func (c *GUIConfiguration) prepare() error {
...
-		newUserId := make([]byte, 64)
+	if len(c.WebauthnUserId) == 0 {
+		// Spec recommends 64 random bytes, we use 32 and hex-encoding inflates to 64
+		newUserId := make([]byte, 32)
 		_, err := rand.Read(newUserId)
 		if err != nil {
 			return err
 		}
-		c.WebauthnUserId = base64.URLEncoding.EncodeToString(newUserId)
+		// Hex-encode the ID so that the raw binary values look nice in config.xml
```
Comment (acolomb) [inline lib/config/guiconfiguration.go]: It's not really about "looking nice" only, but being printable and not conflict with XML syntax.

File: lib/config/guiconfiguration.go [hunk 3]
```diff
@@ -166,13 +166,15 @@ func (c *GUIConfiguration) prepare() error {
 		c.APIKey = rand.String(32)
 	}
 
-	if c.WebauthnUserId == "" {
-		newUserId := make([]byte, 64)
+	if len(c.WebauthnUserId) == 0 {
+		// Spec recommends 64 random bytes, we use 32 and hex-encoding inflates to 64
```
Comment (acolomb) [inline lib/config/guiconfiguration.go]: The amount of entropy is still half of what is recommended, because hex-encoding only "inflates" the number of octets stored, but simultaneously reduces the number of possible code points (`[0-9a-f]`).  So we can't really call it "64 random bytes".  For the same amount of entropy, you'd need hex-encode 64 bytes.

File: lib/api/api.go (hunk 1 above)
Comment (imsodin) [reply lib/api/api.go]: Oh that's indeed odd - usually a nil and zero length slice behave just the same in go. However likely not an issue as we do the same thing everywhere. If you wanted to be sticklerish you could add an `Equal` method, manually compare the two slices and then set them to `nil` for `reflect.DeepEqual`. However imo that's overkill.

File: lib/config/guiconfiguration.go (hunk 3 above)
Comment (imsodin) [reply lib/config/guiconfiguration.go]: The comment doesn't make any claim on entropy?

I also don't find the randomness recommendation in the spec, just this:

> The [user handle](https://www.w3.org/TR/webauthn-2/#user-handle) of the user account entity. A [user handle](https://www.w3.org/TR/webauthn-2/#user-handle) is an opaque [byte sequence](https://infra.spec.whatwg.org/#byte-sequence) with a maximum size of 64 bytes, and is not meant to be displayed to the user.

Neither do I see a need for this to be high entropy - it's not a secret.

Comment (acolomb) [reply lib/config/guiconfiguration.go]: Alright, I assumed the requirement was for actual entropy / "randomness".  In that case it's fine.

Comment (emlun) [reply lib/config/guiconfiguration.go]: Yeah, 32 bytes (256 bits) is more than enough for the value to be probabilistically unique within the application which is all it needs to be. The comment may be a bit misleading, I'll take another pass at it.

File: lib/config/guiconfiguration.go (hunk 2 above)
Comment (emlun) [reply lib/config/guiconfiguration.go]: Thanks, addressed in a2b4c30e827981a08ce000c96b6eff408478ed0d.

File: lib/config/guiconfiguration.go (hunk 3 above)
Comment (emlun) [reply lib/config/guiconfiguration.go]: Comment addressed in a2b4c30e827981a08ce000c96b6eff408478ed0d.

File: lib/api/api.go (hunk 1 above)
Comment (emlun) [reply lib/api/api.go]: Ok, I'll just use `reflect.DeepEqual` for now then. Done in caad2ff7b8ef3fd8a49055650ea0b3b1b444f666.

**Review by acolomb — APPROVED** [review]
Recent changes look OK from my side.  I'll probably need to take another close look within the context of the main PR.

**Comment by emlun** [comment]
Ok, since @imsodin gave the "ok except `cmp.DeepEqual`" I'll go ahead and merge this into https://github.com/syncthing/syncthing/pull/9175, and we can re-review there if needed. :slightly_smiling_face:
